lib/iex/lib/iex/helpers.ex

```diff
@@ -134,12 +134,18 @@ defmodule IEx.Helpers do
 
     modules = Enum.map(erls, fn(source) ->
       {module, binary} = compile_erlang(source)
-      base = source |> Path.basename |> Path.rootname
-      File.write!(Path.join(path, base <> ".beam"), binary)
+      unless path == :in_memory do
+        base = source |> Path.basename |> Path.rootname
+        File.write!(Path.join(path, base <> ".beam"), binary)
+      end
       module
     end)
 
-    modules ++ Kernel.ParallelCompiler.files_to_path(exs, path)
+    modules ++ if path == :in_memory do
+      Kernel.ParallelCompiler.files(exs)
+    else
+      Kernel.ParallelCompiler.files_to_path(exs, path)
+    end
```

Comment:
This looks a bit confusing with the `++` and the `if`. Maybe we can refactor this out to a helper function?

```elixir
modules ++ compile_exs(exs, path)

defp compile_exs(exs, :in_memory), do: Kernel.ParallelCompiler.files(exs)
defp compile_exs(exs, path), do: Kernel.ParallelCompiler.files_to_path(exs, path)
```

? Or maybe at least have

```elixir
compiled_exs =
  if path == :in_memory do
    ...
  end

modules ++ compiled_exs
```

Wdyt?